cpp/src/arrow/type.cc

```diff
@@ -79,6 +105,38 @@ std::shared_ptr<Field> Field::Copy() const {
   return ::arrow::field(name_, type_, nullable_, metadata_);
 }
 
+const std::string& Field::fingerprint() const {
+  if (!fingerprint_computed_) {
+    const auto& type_fingerprint = type_->fingerprint();
+    if (!type_fingerprint.empty()) {
+      std::stringstream ss;
+      ss << 'F';
+      if (nullable_) {
+        ss << 'n';
+      } else {
+        ss << 'N';
+      }
+      ss << name_;
+      ss << '{' << type_fingerprint << '}';
+      if (metadata_) {
+        // Compute metadata fingerprint.  KeyValueMetadata is not immutable,
+        // so we don't cache the result on the metadata instance.
+        ss << "!{";
+        for (int64_t i = 0; i < metadata_->size(); ++i) {
```

Comment:
this seems like it would make the check expensive again if clients made heavy use of metadata.   I wonder if it makes sense to maybe make a first class TypeFingerprint class that can do the comparison incrementally?

Reply:
The fingerprint is cached on the field, so I'm not sure what you mean here.